File: claude-experiments/java-js/benchmarks/rust/src/benchmark_real_world.rs
/*!
 * Real-World JavaScript Library Benchmarks (Rust)
 */

use std::env;
use std::fs;
use std::time::Instant;
use std::sync::Arc;

// OXC imports
use oxc_allocator::Allocator;
use oxc_parser::Parser as OxcParser;
use oxc_span::SourceType;

// SWC imports
use swc_common::SourceMap;
use swc_ecma_parser::{Parser as SwcParser, StringInput, Syntax, EsSyntax};

const LIBS_DIR: &str = "../real-world-libs/";

struct BenchmarkResult {
    name: String,
    library: String,
    avg_millis: f64,
    throughput_kb_per_ms: f64,
}

const DEFAULT_WARMUP_ITERATIONS: usize = 5;
const DEFAULT_MEASUREMENT_ITERATIONS: usize = 10;

fn benchmark<F>(name: &str, library: &str, code: &str, mut parse_fn: F, warmup_iterations: usize, iterations: usize) -> BenchmarkResult
where
    F: FnMut(&str),
{
    // Warmup
    for _ in 0..warmup_iterations {
        parse_fn(code);
    }

    // Measure
    let start = Instant::now();
    for _ in 0..iterations {
        parse_fn(code);
    }
    let elapsed = start.elapsed();

    // Use as_secs_f64() for precise sub-millisecond measurements
    let avg_millis = (elapsed.as_secs_f64() * 1000.0) / iterations as f64;
    let size_kb = code.len() as f64 / 1024.0;
    let throughput_kb_per_ms = size_kb / avg_millis;

    BenchmarkResult {
        name: name.to_string(),
        library: library.to_string(),
        avg_millis,
        throughput_kb_per_ms,
    }
}

fn parse_with_oxc(code: &str) {
    let allocator = Allocator::default();
    let source_type = SourceType::default();
    let _ = OxcParser::new(&allocator, code, source_type).parse();
}

fn parse_with_swc(code: &str) {
    let cm = Arc::new(SourceMap::default());
    let fm = cm.new_source_file(swc_common::FileName::Anon.into(), code.to_string());
    let input = StringInput::from(&*fm);

    let syntax = Syntax::Es(EsSyntax {
        jsx: false,
        fn_bind: false,
        decorators: false,
        decorators_before_export: false,
        export_default_from: false,
        import_attributes: true,
        allow_super_outside_method: false,
        allow_return_outside_function: false,
        auto_accessors: false,
        explicit_resource_management: false,
    });

    let mut parser = SwcParser::new(syntax, input, None);
    let _ = parser.parse_module();
}

fn run_benchmark_suite(library_name: &str, filename: &str, warmup_iterations: usize, iterations: usize) {
    let path = format!("{}{}", LIBS_DIR, filename);
    let code = match fs::read_to_string(&path) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("Failed to read {}: {}", path, e);
            return;
        }
    };

    let size_kb = code.len() as f64 / 1024.0;

    println!("\n{}", "=".repeat(80));
    println!("Library: {}", library_name);
    println!("Size: {:.1} KB ({} bytes)", size_kb, code.len());
    println!("{}\n", "=".repeat(80));

    let mut results = vec![
        benchmark("OXC (Rust)", library_name, &code, |c| parse_with_oxc(c), warmup_iterations, iterations),
        benchmark("SWC (Rust)", library_name, &code, |c| parse_with_swc(c), warmup_iterations, iterations),
    ];

    results.sort_by(|a, b| a.avg_millis.partial_cmp(&b.avg_millis).unwrap());

    println!("Results:");
    println!("{}", "-".repeat(80));
    println!("{:<20} | {:>15} | {:>15} | {:>20}", "Parser", "Avg Time (ms)", "vs Fastest", "Throughput (KB/ms)");
    println!("{}", "-".repeat(80));

    let fastest = results[0].avg_millis;

    for (i, result) in results.iter().enumerate() {
        let ratio = result.avg_millis / fastest;
        let indicator = match i {
            0 => "🥇",
            1 => "🥈",
            _ => "  ",
        };

        println!(
            "{} {:<18} | {:>15.3} | {:>15.2}x | {:>20.1}",
            indicator,
            result.name,
            result.avg_millis,
            ratio,
            result.throughput_kb_per_ms
        );
    }
}

fn main() {
    // Parse command line arguments: benchmark-real-world [warmup] [measurement]
    let args: Vec<String> = env::args().collect();
    let warmup_iterations = args.get(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(DEFAULT_WARMUP_ITERATIONS);
    let measurement_iterations = args.get(2)
        .and_then(|s| s.parse().ok())
        .unwrap_or(DEFAULT_MEASUREMENT_ITERATIONS);

    println!("\n{}", "═".repeat(80));
    println!("Real-World JavaScript Library Benchmarks (Rust)");
    println!("{}", "═".repeat(80));
    println!("Warmup iterations: {}", warmup_iterations);
    println!("Measurement iterations: {}", measurement_iterations);
    println!("{}", "═".repeat(80));

    run_benchmark_suite("React", "react.production.min.js", warmup_iterations, measurement_iterations);
    run_benchmark_suite("Vue 3", "vue.global.prod.js", warmup_iterations, measurement_iterations);
    run_benchmark_suite("React DOM", "react-dom.production.min.js", warmup_iterations, measurement_iterations);
    run_benchmark_suite("Lodash", "lodash.js", warmup_iterations, measurement_iterations);
    run_benchmark_suite("Three.js", "three.js", warmup_iterations, measurement_iterations);
    run_benchmark_suite("TypeScript Compiler", "typescript.js", warmup_iterations, measurement_iterations);

    println!("\n{}", "═".repeat(80));
    println!("Benchmark complete!");
    println!("{}", "═".repeat(80));
}

File: claude-experiments/java-js/benchmarks/javascript/benchmark-real-world.js
#!/usr/bin/env node

/**
 * Real-World JavaScript Library Benchmarks
 *
 * Tests parser performance on actual production libraries:
 * - TypeScript compiler (8.6 MB)
 * - Three.js (1.3 MB)
 * - Lodash (531 KB)
 * - Vue 3 (130 KB)
 * - React DOM (129 KB)
 */

import * as acorn from 'acorn';
import { parse as babelParse } from '@babel/parser';
import { parseScript as meriyahParse } from 'meriyah';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const LIBS_DIR = join(__dirname, '..', 'real-world-libs');

// Parse command line arguments: node benchmark-real-world.js [warmup] [measurement]
const args = process.argv.slice(2);
const WARMUP_ITERATIONS = args[0] ? parseInt(args[0], 10) : 5;
const MEASUREMENT_ITERATIONS = args[1] ? parseInt(args[1], 10) : 10;

/**
 * Load a library file
 */
function loadLibrary(filename) {
    const path = join(LIBS_DIR, filename);
    const code = readFileSync(path, 'utf-8');
    return { filename, code, size: code.length };
}

/**
 * High-precision benchmark function
 */
function benchmark(name, fn, iterations) {
    const start = process.hrtime.bigint();

    for (let i = 0; i < iterations; i++) {
        fn();
    }

    const end = process.hrtime.bigint();
    const totalNanos = Number(end - start);
    const avgMillis = (totalNanos / iterations) / 1_000_000;

    return {
        name,
        iterations,
        totalNanos,
        avgMillis: avgMillis.toFixed(3)
    };
}

/**
 * Run benchmark suite for a specific library
 */
function runBenchmarkSuite(library) {
    console.log(`\n${'='.repeat(80)}`);
    console.log(`Library: ${library.filename}`);
    console.log(`Size: ${(library.size / 1024).toFixed(1)} KB (${library.size.toLocaleString()} bytes)`);
    console.log(`${'='.repeat(80)}\n`);

    const parsers = [
        {
            name: 'Acorn',
            parse: () => acorn.parse(library.code, { ecmaVersion: 'latest' })
        },
        {
            name: '@babel/parser',
            parse: () => babelParse(library.code)
        },
        {
            name: 'Meriyah',
            parse: () => meriyahParse(library.code, { next: true })
        }
    ];

    const results = [];

    for (const parser of parsers) {
        console.log(`Warming up ${parser.name}...`);

        // Warmup
        try {
            for (let i = 0; i < WARMUP_ITERATIONS; i++) {
                parser.parse();
            }
        } catch (error) {
            console.log(`  ⚠️  ${parser.name} failed: ${error.message}`);
            continue;
        }

        // Benchmark
        console.log(`Benchmarking ${parser.name}...`);
        try {
            const result = benchmark(parser.name, parser.parse, MEASUREMENT_ITERATIONS);
            results.push(result);
        } catch (error) {
            console.log(`  ⚠️  ${parser.name} failed: ${error.message}`);
        }
    }

    if (results.length === 0) {
        console.log('\n⚠️  All parsers failed for this library\n');
        return null;
    }

    // Sort by performance (fastest first)
    results.sort((a, b) => parseFloat(a.avgMillis) - parseFloat(b.avgMillis));

    // Print results
    console.log('\nResults (sorted by performance):');
    console.log('-'.repeat(80));
    console.log(`${'Parser'.padEnd(20)} | ${'Avg Time (ms)'.padStart(15)} | ${'vs Fastest'.padStart(15)} | ${'Throughput'.padStart(15)}`);
    console.log('-'.repeat(80));

    const fastest = parseFloat(results[0].avgMillis);

    for (const result of results) {
        const avg = parseFloat(result.avgMillis);
        const ratio = (avg / fastest).toFixed(2);
        const throughput = (library.size / avg / 1024).toFixed(1); // KB/ms
        const indicator = result === results[0] ? '🥇' :
                         result === results[1] ? '🥈' :
                         result === results[2] ? '🥉' : '  ';

        console.log(
            `${indicator} ${result.name.padEnd(18)} | ${result.avgMillis.padStart(15)} | ${ratio.padStart(15)}x | ${throughput.padStart(12)} KB/ms`
        );
    }

    return results;
}

/**
 * Main benchmark runner
 */
async function main() {
    console.log('\n' + '═'.repeat(80));
    console.log('Real-World JavaScript Library Parser Benchmarks');
    console.log('═'.repeat(80));
    console.log(`Warmup iterations: ${WARMUP_ITERATIONS}`);
    console.log(`Measurement iterations: ${MEASUREMENT_ITERATIONS}`);
    console.log(`Node.js version: ${process.version}`);
    console.log('═'.repeat(80));

    const libraries = [
        loadLibrary('react.production.min.js'),
        loadLibrary('vue.global.prod.js'),
        loadLibrary('react-dom.production.min.js'),
        loadLibrary('lodash.js'),
        loadLibrary('three.js'),
        loadLibrary('typescript.js'),
    ];

    const allResults = {};

    for (const library of libraries) {
        const results = runBenchmarkSuite(library);
        if (results) {
            allResults[library.filename] = results;
        }
    }

    // Summary
    console.log('\n' + '═'.repeat(80));
    console.log('SUMMARY');
    console.log('═'.repeat(80));

    for (const [libName, results] of Object.entries(allResults)) {
        if (results.length > 0) {
            console.log(`\n${libName}:`);
            console.log(`  🥇 ${results[0].name}: ${results[0].avgMillis} ms`);
        }
    }

    console.log('\n' + '═'.repeat(80));
}

main().catch(console.error);
